File: algorithms/minStack/minStack.go
package minStack

type stack struct {
	s   []int
	min int //index of min
}

func (p *stack) push(n int) {
	if len(p.s) == 0 {
		p.min = 0
	} else if n < p.s[p.min] {
		p.min = len(p.s)
	}
	p.s = append(p.s, n)
}

func (p *stack) pop() {
	if len(p.s) == 0 {
		return
	}
	p.s = p.s[:len(p.s)-1]
	if p.min == len(p.s) {
		if len(p.s) == 1 {
			p.min = 0
		} else {
			m := p.s[0]
			for i := 1; i < len(p.s); i++ {
				if p.s[i] < m {
					m = p.s[i]
					p.min = i
				}
			}
		}
	}
}

func (p *stack) top() int {
	if len(p.s) == 0 {
		return 0
	}
	return p.s[len(p.s)-1]
}

func (p *stack) getMin() int {
	if len(p.s) == 0 {
		return 0
	}
	return p.s[p.min]
}

File: algorithms/java/src/minStack/MinStack.java
// Source : https://leetcode.com/problems/min-stack/description/
// Author : Tianming Cao
// Date   : 2018-02-02

/**********************************************************************************
 * Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
 *
 *	push(x) -- Push element x onto stack.
 *	pop() -- Removes the element on top of the stack.
 *	top() -- Get the top element.
 *	getMin() -- Retrieve the minimum element in the stack.
 * Example:
 *	MinStack minStack = new MinStack();
 *	minStack.push(-2);
 *	minStack.push(0);
 *	minStack.push(-3);
 *	minStack.getMin();   --> Returns -3.
 *	minStack.pop();
 *	minStack.top();      --> Returns 0.
 *	minStack.getMin();   --> Returns -2.
 * 
 **********************************************************************************/
package minStack;

import java.util.Stack;

public class MinStack {
	public Stack<Integer> mainStack;
	/**
	 * Call an extra stack named assistStack to store the min value.
	 * While we doing push operation, compare x with the top of assistStack and push the smaller value into assistStack.
	 * The other operations pop,top and getMin is very simple.
	 **/
	public Stack<Integer> assistStack;
	public MinStack() {
		mainStack = new Stack<Integer>();
		assistStack = new Stack<Integer>();
	}
	public void push(int x) {
		mainStack.push(x);
		if (assistStack.isEmpty()) {
			assistStack.push(x);
		} else {
			assistStack.push(Math.min(x, getMin()));
		}
	}

	public void pop() {
		mainStack.pop();
		assistStack.pop();
	}

	public int top() {
		return mainStack.peek();
	}

	public int getMin() {
		return assistStack.peek();
	}
}
